### ISSUE TITLE
Better structure for APISummarizer

### ISSUE THREAD

**Issue by HosseinYousefi** [dart-lang/jnigen#157]
* Currently we have two parsers, one for sources and one for bytecode. Using metadata, these two will almost be the same with the exception of docs which could be fetched in case they're available. 
* The JSON export is very `string`y. It would be better for classes to have unique ids and actually pass in the full AST for an exact representation for the codegen. The codegen, uses a map of `String` to `ClassDecl` to resolve the classes. We need to have the same "resolving" logic for type bounds, generic type variables, etc. Using string names is not the way to go for all of these use-cases.
* I find the way `TypeUsage` has been declared to be confusing.

TL;DR: We would want to pass a tree instead of JSON objects and lookup values from a map (even if the format is still JSON).

**Comment by dcharkes** [comment]
I believe it was string-y because identifiers are uniquely prefixed.
If references are resolved to the actual definitions, it will be a graph, not a tree, which is harder to serialize.

I have no objections against making the serialized format better.

Side note: because the ApiSummarizer is a separate process (rather than using jnigen itself to directly use the java packages for parsing) enables us to have different licensing for this separate program. Licenses are contagious when "linking" things together through FFI, but not when "invoking" a separate program through `Process.run`. So we can't use jnigen on the Java libraries and interact directly with their output without serialization step in between. (Which is what we do in ffigen, and what I had hoped we could do.)

@mahesh-hegde any thoughts based on your experience writing the summarizer?

**Comment by mahesh-hegde** [comment]
> The JSON export is very stringy. It would be better for classes to have unique ids and actually pass in the full AST for an exact representation for the codegen.

Initially we only had declared classes. So having string keys referring to other classes made sense.

Also, this representation makes no difference between class in current set of classes vs some external class (which may be provided through an import map for example..). In my design, resolver object is supposed to take care of that.

> TypeUsage

It is basically serialised form of `TypeMirror` in doclet API. It's to differentiate between __declaration__ of a type vs __usage__, latter can have type parameters etc..

I have tried to mirror doclet API as close as possible. So ClassDecl is more analogue to `DeclaredType` than `TypeMirror`. A declared type is always in scope, but a type parameter or wildcard is only stored in a type usage and it's meaning depends on scope.

Re: comment. It's a fairly minor implementation detail. It's still a inheritance hierarchy where `PrimitiveType` etc.. extend `TypeUsage`. but deserializing it from JSON needs a hint of which type to deserialize into.

You can indeed shorten lot of these outputs from the summarizer. I have just tried to keep uniformity among all `TypeUsage` subclasses.

**Comment by HosseinYousefi** [comment]
Decided not to pursue this for now.
